File: src/lib/scanner.ts
import fs from 'fs/promises';
import path from 'path';
import os from 'os';
import { glob } from 'glob';
import { ScannedContext, Config, UnifiedRegistry, ContextPathConfig } from './types.js';
import { DEFAULT_PATTERNS } from './config.js';

// New 3-level constants
const CTX_DIR = '.ctx';
const CONTEXTS_DIR = 'contexts';
const REGISTRY_FILE = 'registry.yaml';

/** Get global ctx directory (uses current HOME, not cached) */
function getGlobalCtxDir(): string {
  return path.join(os.homedir(), CTX_DIR);
}

/** Default context paths when settings not configured */
const DEFAULT_CONTEXT_PATHS: ContextPathConfig[] = [
  { path: '**/*.ctx.md', purpose: 'Bound contexts next to code' },
  { path: '.ctx/contexts/**/*.md', purpose: 'Centralized project contexts' },
];

/**
 * Scan for local context files based on config patterns
 */
export async function scanLocalContexts(
  projectRoot: string,
  config: Config
): Promise<ScannedContext[]> {
  const contexts: ScannedContext[] = [];

  // Support both string and array patterns
  const patterns = Array.isArray(config.local.patterns)
    ? config.local.patterns
    : [config.local.patterns];

  for (const pattern of patterns) {
    const files = await glob(pattern, {
      cwd: projectRoot,
      ignore: [
        ...config.local.ignore,
        `${config.global.directory}/**`,
      ],
      absolute: false,
    });

    for (const file of files) {
      const absolutePath = path.join(projectRoot, file);
      try {
        const content = await fs.readFile(absolutePath, 'utf-8');
        contexts.push({
          contextPath: absolutePath,
          relativePath: file,
          content,
        });
      } catch (error) {
        console.error(`Warning: Failed to read ${file}: ${error}`);
      }
    }
  }
 
  return contexts;
}

/**
 * Scan for global context files based on config patterns
 */
export async function scanGlobalContexts(
  projectRoot: string,
  config: Config
): Promise<ScannedContext[]> {
  const globalDir = path.join(projectRoot, config.global.directory);

  // Check if global directory exists
  try {
    await fs.access(globalDir);
  } catch {
    return []; // global directory doesn't exist
  }

  // Build pattern with config directory
  const pattern = `${config.global.directory}/${config.global.patterns}`;

  const files = await glob(pattern, {
    cwd: projectRoot,
    ignore: [
      ...config.global.ignore.map(p => `${config.global.directory}/${p}`),
    ],
    absolute: false,
  });

  const contexts: ScannedContext[] = [];

  for (const file of files) {
    const absolutePath = path.join(projectRoot, file);

    try {
      const content = await fs.readFile(absolutePath, 'utf-8');
      contexts.push({
        contextPath: absolutePath,
        relativePath: file,
        content,
      });
    } catch (error) {
      console.error(`Warning: Failed to read ${file}: ${error}`);
    }
  }

  return contexts;
}

/**
 * Extract folder from global context path
 * Example: ctx/rules/input-handling.md -> 'rules'
 * Example: ctx/overview.md -> null
 */
export function extractFolder(relativePath: string, globalDir: string): string | null {
  // Remove global directory prefix (e.g., 'ctx/' or 'docs/')
  const pattern = new RegExp(`^${globalDir.replace(/[/\\]/g, '[/\\\\]')}[/\\\\]`);
  const withoutDir = relativePath.replace(pattern, '');

  // Get first directory
  const parts = withoutDir.split(/[/\\]/);

  // If only one part, it's a root file
  if (parts.length === 1) {
    return null;
  }

  // Otherwise, return the folder name
  return parts[0];
}

// ===== New 3-Level Scanner Functions =====

/**
 * Get context paths from registry settings
 * Falls back to default paths if settings not configured
 */
async function getContextPathsFromRegistry(registryPath: string): Promise<ContextPathConfig[]> {
  try {
    const content = await fs.readFile(registryPath, 'utf-8');
    const { parse } = await import('yaml');
    const registry = parse(content) as UnifiedRegistry;
    if (registry.settings?.context_paths && registry.settings.context_paths.length > 0) {
      return registry.settings.context_paths;
    }
  } catch {
    // Registry doesn't exist or is invalid
  }
  return DEFAULT_CONTEXT_PATHS;
}

/**
 * Scan ALL project context files based on context_paths settings.
 * Default patterns include bound contexts and centralized contexts.
 *
 * All are project contexts - distinction is by target field presence, not location.
 */
export async function scanProjectContexts(
  projectRoot: string
): Promise<ScannedContext[]> {
  const contexts: ScannedContext[] = [];
  const seenPaths = new Set<string>();

  const registryPath = path.join(projectRoot, CTX_DIR, REGISTRY_FILE);
  const contextPaths = await getContextPathsFromRegistry(registryPath);

  for (const cp of contextPaths) {
    // Build ignore list: exclude .ctx/** unless this pattern explicitly targets .ctx/
    const ignore = cp.path.startsWith('.ctx/')
      ? DEFAULT_PATTERNS.ignore.filter(p => p !== '.ctx/**')
      : DEFAULT_PATTERNS.ignore;

    const files = await glob(cp.path, {
      cwd: projectRoot,
      ignore,
      absolute: false,
    });

    for (const file of files) {
      if (seenPaths.has(file)) continue;
      seenPaths.add(file);

      const absolutePath = path.join(projectRoot, file);
      try {
        const content = await fs.readFile(absolutePath, 'utf-8');
        contexts.push({
          contextPath: absolutePath,
          relativePath: file,
          content,
        });
      } catch (error) {
        console.error(`Warning: Failed to read ${file}: ${error}`);
      }
    }
  }

  return contexts;
}

/**
 * Scan for global context files based on settings.context_paths in ~/.ctx/
 */
export async function scanGlobalCtxContexts(): Promise<ScannedContext[]> {
  const globalCtxDir = getGlobalCtxDir();
  const registryPath = path.join(globalCtxDir, REGISTRY_FILE);
  const contextPaths = await getContextPathsFromRegistry(registryPath);

  const contexts: ScannedContext[] = [];
  const seenPaths = new Set<string>();

  for (const cp of contextPaths) {
    const files = await glob(cp.path, {
      cwd: globalCtxDir,
      absolute: false,
    });

    for (const file of files) {
      if (seenPaths.has(file)) continue;
      seenPaths.add(file);

      const absolutePath = path.join(globalCtxDir, file);
      try {
        const content = await fs.readFile(absolutePath, 'utf-8');
        contexts.push({
          contextPath: absolutePath,
          relativePath: file,
          content,
        });
      } catch (error) {
        console.error(`Warning: Failed to read ${file}: ${error}`);
      }
    }
  }

  return contexts;
}

File: src/scanner.rs
use crate::parser;
use crate::project;
use crate::types::{ContextNode, CtxConfig};
use std::collections::HashMap;
use std::path::Path;
use walkdir::WalkDir;
use globset::{Glob, GlobSetBuilder};

/// Scan the project for all valid context nodes
pub fn scan_nodes(project_root: &Path, config: &CtxConfig) -> (Vec<ContextNode>, Vec<String>) {
    let mut nodes = Vec::new();
    let mut warnings = Vec::new();

    // Build ignore globset
    let mut builder = GlobSetBuilder::new();
    for pattern in &config.ignore {
        if let Ok(glob) = Glob::new(pattern) {
            builder.add(glob);
        }
    }
    let ignore_set = builder.build().unwrap_or_else(|_| GlobSetBuilder::new().build().unwrap());

    for entry in WalkDir::new(project_root)
        .into_iter()
        .filter_map(|e| e.ok())
    {
        let path = entry.path();

        // Only process .md files
        if path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }

        let rel = project::relative_path(path, project_root);

        // Check ignore patterns
        if ignore_set.is_match(&rel) {
            continue;
        }

        // Read and parse
        let content = match std::fs::read_to_string(path) {
            Ok(c) => c,
            Err(_) => continue,
        };

        match parser::parse_context_file(&content, &rel) {
            Ok(Some(node)) => nodes.push(node),
            Ok(None) => {}, // Not a ctx node, skip silently
            Err(e) => warnings.push(e),
        }
    }

    (nodes, warnings)
}

/// Build a name→node map, detecting duplicates
pub fn build_node_map(nodes: Vec<ContextNode>) -> (HashMap<String, ContextNode>, Vec<(String, Vec<String>)>) {
    let mut map: HashMap<String, ContextNode> = HashMap::new();
    let mut all_paths: HashMap<String, Vec<String>> = HashMap::new();

    // Sort nodes by path for deterministic first-found behavior
    let mut sorted_nodes = nodes;
    sorted_nodes.sort_by(|a, b| a.path.cmp(&b.path));

    for node in sorted_nodes {
        all_paths
            .entry(node.name.clone())
            .or_default()
            .push(node.path.clone());

        map.entry(node.name.clone()).or_insert(node);
    }

    let duplicates: Vec<(String, Vec<String>)> = all_paths
        .into_iter()
        .filter(|(_, paths)| paths.len() > 1)
        .collect();

    (map, duplicates)
}
